### PR TITLE
Fix issue 1808

### PR DESCRIPTION
**Pull request by edmondchuc**
# Summary of changes

Fixes https://github.com/RDFLib/rdflib/issues/1808

Essentially we are reverting back to returning a `URIRef` in the `URIRef.skolemize()` method and changing how `Graph.de_skolemize()` method converts subjects and objects of a statement to a `BNode`.

I think this change is much safer than changing what is returned by `URIRef.skolemize()` and potentially affecting other parts of the library. I've had a search through the codebase and nothing else touches the `Graph.de_skolemize()`.

# Checklist

- [x] Checked that there aren't other open pull requests for
  the same change.
- [x] Added tests for any changes that have a runtime impact.
- [x] Checked that all tests and type checking passes.
- For changes that have a potential impact on users of this project:
  - [x] Updated relevant documentation to avoid inaccuracies.
  - [x] Considered adding additional documentation.
  - [x] Considered adding an example in `./examples` for new features.
  - [x] Considered updating our changelog (`CHANGELOG.md`).
- [x] Considered granting [push permissions to the PR branch](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork),
  so maintainers can fix minor issues and keep your PR up to date.

### CODE REVIEW

**Comment by aucampia** [comment]
@edmondchuc - pre commit errors can be fixed by making a comment "pre-commit.ci autofix"

image

**Comment by nicholascar** [comment]
pre-commit.ci autofix

**Comment by edmondchuc** [comment]
Thanks for the tip @aucampia and thanks @nicholascar for commenting.

I'm hoping this PR fixes both https://github.com/RDFLib/rdflib/issues/1808 and https://github.com/RDFLib/rdflib/issues/1404 without introducing further side effects.

**Review by gjhiggins — APPROVED** [review]
Nice work.

**Comment by aucampia** [comment]
Will review in about an hour, just fixing some other issues.

**Review by gjhiggins — APPROVED** [review]

File: rdflib/graph.py [hunk 1]
```diff
@@ -1510,10 +1509,17 @@ def do_de_skolemize(uriref, t):
...
             (s, p, o) = t
-            if isinstance(s, Genid):
-                s = s.de_skolemize()
-            if isinstance(o, Genid):
-                o = o.de_skolemize()
+
+            if Genid._is_external_skolem(s):
+                s = Genid(s).de_skolemize()
+            elif RDFLibGenid._is_rdflib_skolem(s):
+                s = RDFLibGenid(s).de_skolemize()
```
Comment (aucampia) [inline rdflib/graph.py:1516]: I'm going to add tests to confirm, but I think the second branch here `elif RDFLibGenid._is_rdflib_skolem(s)` may never execute, as the condition for `Genid._is_external_skolem(s)` will always be true if `RDFLibGenid._is_rdflib_skolem(s)` is true.

File: test/test_issues/test_issue1808.py [hunk 2]
```diff
@@ -0,0 +1,44 @@
+from rdflib import Graph
+from rdflib.term import BNode, URIRef, rdflib_skolem_genid
```
Comment (aucampia) [inline test/test_issues/test_issue1808.py:2]: Ideally there should be tests for this in a more general test suite, as I think in the long term we have to eliminate all the issue specific tests.

Comment (aucampia) [reply test/test_issues/test_issue1808.py:2]: Fine for now though.

File: rdflib/graph.py (hunk 1 above)
Comment (aucampia) [reply rdflib/graph.py:1516]: Actually I may only get to this tomorrow, there are quite a backlog of issues I'm trying to fix right now, but I before we have a test specifically verifying this I would not be comfortable with merging.

Comment (aucampia) [reply rdflib/graph.py:1516]: Feel free to add it in the meantime, I will let you know when I start working on this.

Comment (edmondchuc) [reply rdflib/graph.py:1516]: I think you are right. We should do the `RDFLibGenid._is_rdflib_skolem` check before `Genid._is_external_skolem`. I'll add a test to confirm.

**Comment by edmondchuc** [comment]
pre-commit.ci autofix

File: rdflib/graph.py [hunk 3]
```diff
@@ -1510,10 +1509,19 @@ def do_de_skolemize(uriref, t):
 
         def do_de_skolemize2(t):
             (s, p, o) = t
-            if isinstance(s, Genid):
-                s = s.de_skolemize()
-            if isinstance(o, Genid):
-                o = o.de_skolemize()
+
+            if RDFLibGenid._is_rdflib_skolem(s):
+                print('rdflibgen')
```
Comment (aucampia) [inline rdflib/graph.py:1514]: I think you forgot this here:
```suggestion
```

File: rdflib/graph.py [hunk 4]
```diff
@@ -1510,10 +1509,19 @@ def do_de_skolemize(uriref, t):
...
+                s = RDFLibGenid(s).de_skolemize()
+            elif Genid._is_external_skolem(s):
+                print('genid')
```
Comment (aucampia) [inline rdflib/graph.py:1517]: I think you forgot this here:
```suggestion
```

File: rdflib/graph.py (hunk 3 above)
Comment (edmondchuc) [reply rdflib/graph.py]: Whoops, thanks!

**Comment by aucampia** [comment]
@edmondchuc I have added some more tests now - and I'm fairly comfortable with the change after adding them - hope they are okay with you.

**Review by aucampia — APPROVED** [review]
Fix looks good, thanks @edmondchuc

**Comment by aucampia** [comment]
Planning to merge on or before 2022-04-17 if there is no further feedback.

**Comment by edmondchuc** [comment]
> @edmondchuc I have added some more tests now - and I'm fairly comfortable with the change after adding them - hope they are okay with you.

Thanks for adding the additional tests @aucampia.

**Comment by aucampia** [comment]
Won't merge before #1838

**Pull request merged**